### PR TITLE
Refactor work alone group to update repo permission once and outside the transaction

### PR DESCRIPTION
**Pull request by adisandro**
This applies the same principles of #3296 for the regeneration of the repo auth file.
In addition, refactors the logic around a weird corner case.

### CODE REVIEW

File: app/models/student.rb [hunk 1]
```diff
@@ -141,45 +138,33 @@ def create_group_for_working_alone_student(aid)
         end
       end
 
-      @grouping.group = @group
-      begin
-        unless @grouping.save
-          m_logger = MarkusLogger.instance
-          m_logger.log("Could not create a grouping for Student '#{user_name}'"\
+      empty_grouping = Grouping.find_by(assignment_id: aid, group_id: @group.id)
```
Comment (david-yz-liu) [inline app/models/student.rb]: I looked at the logic and the old issue. This seems unnecessary (at least now, maybe useful before). If a grouping already exists, it should still be the case that we can add the student to the grouping without destroying and recreating it.

File: app/controllers/assignments_controller.rb [hunk 2]
```diff
@@ -104,15 +104,7 @@ def student_interface
...
-          # when saving a grouping throws an exception
-          unless @student.create_group_for_working_alone_student(@assignment.id)
-            # if create_group_for_working_alone_student returned false then the student
-            # must have an ( empty ) existing grouping that he is not a member of.
-            # we must delete this grouping for the transaction to succeed.
-            Grouping.find_by_group_id_and_assignment_id( Group.find_by_group_name(@student.user_name), @assignment.id).destroy
-          end
+          @student.create_group_for_working_alone_student(@assignment.id)
         rescue RuntimeError => @error
           flash_message(:error, 'Error')
```
Comment (david-yz-liu) [inline app/controllers/assignments_controller.rb]: Do you mind changing this to use the error's message? (And I think you can change `@error` to `error`). I know it's not exactly relevant, but I think it's a related-enough improvement.

File: app/models/student.rb [hunk 3]
```diff
@@ -141,45 +138,33 @@ def create_group_for_working_alone_student(aid)
...
           ' come back to this page, you should inform the course instructor.'
-        end
-      # This exception will only be thrown when we try to save to a grouping that already exists
-      rescue ActiveRecord::RecordNotUnique
-        # transaction has failed, so quit it
-        return false
-      end
-
-      # We give students the tokens for the test framework
-      if @assignment.enable_test
```
Comment (david-yz-liu) [inline app/models/student.rb]: Here I'm just checking that this happens somewhere else? Or is unnecessary?

File: app/models/student.rb [hunk 4]
```diff
@@ -141,45 +138,33 @@ def create_group_for_working_alone_student(aid)
...
-      @grouping.update_repository_permissions
-      # Add permissions for TAs and admins.  This completely rerwrites the auth file
-      # but that shouldn't be a big deal in this case.
-      @group.set_repo_permissions
+      self.destroy_all_pending_memberships(aid)
     end
-    return true
+    Repository.get_class.__set_all_permissions
+
+    true
```
Comment (david-yz-liu) [inline app/models/student.rb]: I'd prefer if you remove the return value altogether now. There are two tests in `student_test` that you'll need to modify, but that shouldn't be too long.

File: app/models/student.rb (hunk 1 above)
Comment (adisandro) [reply app/models/student.rb]: There is a unique index in the database on (assignement_id + group_id) in the groupings table.
I was able to reproduce the issue: you create your group as a student, then log in as the instructor and remove the student membership from the group (without deleting the group), then try to create it again as a student.

File: app/models/student.rb (hunk 3 above)
Comment (adisandro) [reply app/models/student.rb]: It's unnecesary, tokens are given at test request time if needed.
(but thinking about it, let me double check the interface instead, whether it hides the button without tokens)

Comment (adisandro) [reply app/models/student.rb]: All good!

**Comment by david-yz-liu** [comment]
@adisandro since you're still working on this: I went through and deleted the Hound comments I don't care about and will disable, but there were two that I liked and I ask that you change. Thanks!

**Comment by adisandro** [comment]
@david-yz-liu I was done actually, but I'll look into these Hound comments.

File: app/models/student.rb (hunk 1 above)
Comment (david-yz-liu) [reply app/models/student.rb]: @adisandro weren't you going to replace this with a `find_or_create_by`?

Comment (adisandro) [reply app/models/student.rb]: Shoot I confused this pull request with the other about git.

File: app/models/student.rb [hunk 5]
```diff
@@ -141,45 +138,31 @@ def create_group_for_working_alone_student(aid)
         end
       end
 
-      @grouping.group = @group
-      begin
-        unless @grouping.save
-          m_logger = MarkusLogger.instance
-          m_logger.log("Could not create a grouping for Student '#{user_name}'"\
+      empty_grouping = Grouping.find_by(assignment_id: aid, group_id: @group.id)
+      unless empty_grouping.nil?
```
[the remark this answers is not in the document]
Comment (adisandro) [reply app/models/student.rb]: I'd like to, if we weren't running ruby 2.2 in production

**Pull request merged**